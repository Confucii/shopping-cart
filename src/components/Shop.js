import "../styles/Shop.css";

export default function Shop() {
  return <div className="Shop">Shop</div>;
}
