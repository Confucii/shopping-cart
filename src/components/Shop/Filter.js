import "./styles/Filter.css";

export default function Filter({ setFilter, search, setSearch }) {
  function toggleFilter(e) {
    if (e.currentTarget.classList.contains("chosen")) {
      e.currentTarget.classList.remove("chosen");
      setFilter("");
    } else {
      document.querySelector(".chosen")?.classList.remove("chosen");
      e.currentTarget.classList.add("chosen");
      setFilter(e.currentTarget.textContent.toLowerCase());
    }
  }

  function handleSearchChange(e) {
    setSearch(e.currentTarget.value);
  }

  return (
    <div className="Filter">
      <p className="search-name">Search</p>
      <input
        type="text"
        name="search-input"
        id="search-input"
        onChange={handleSearchChange}
        value={search}
      />
      <p className="filter-name">Filter</p>
      <ul className="filter-list">
        <li className="filter-item" onClick={toggleFilter}>
          Red
        </li>
        <li className="filter-item" onClick={toggleFilter}>
          Pink
        </li>
        <li className="filter-item" onClick={toggleFilter}>
          Black
        </li>
        <li className="filter-item" onClick={toggleFilter}>
          Gold
        </li>
      </ul>
    </div>
  );
}
