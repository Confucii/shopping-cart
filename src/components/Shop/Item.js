import { useNavigate } from "react-router-dom";
import "./styles/Item.css";

export default function Item({ item, addItem }) {
  const navigator = useNavigate();

  function reroute(e) {
    navigator(`${e.currentTarget.dataset.id}`);
  }

  function addToCart(e) {
    addItem(item);
    e.stopPropagation();
  }

  return (
    <div data-id={item.id} className="Item" onClick={reroute}>
      <img className="item-img" src={item.image} alt={item.name} />
      <p>Model: {item.name}</p>
      <p>Price: {item.price} credits</p>
      <button className="item-btn" onClick={addToCart}>
        Add to cart
      </button>
    </div>
  );
}
