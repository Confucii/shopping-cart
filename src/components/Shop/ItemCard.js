import { useParams } from "react-router-dom";
import "./styles/ItemCard.css";
import catalogue from "../shopping";

export default function ItemCard({ addItem }) {
  const params = useParams();
  const item = catalogue[params.id];

  function addToCart() {
    addItem(item);
  }

  return (
    <div className="ItemCard">
      <div className="item-card-description">
        <p>Model name: {item.name}</p>
        <p>Color: {item.type.slice(0, 1).toUpperCase() + item.type.slice(1)}</p>
        <p>Description: {item.description}</p>
        <p>Price: {item.price}</p>
        <button className="item-card-btn" onClick={addToCart}>
          Add to cart
        </button>
      </div>
      <img className="item-card-image" src={item.image} alt={item.name} />
    </div>
  );
}
