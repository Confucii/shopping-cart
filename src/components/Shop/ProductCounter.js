import "./styles/ProductCounter.css";

export default function ProductCounter({ item, changeItemCount }) {
  function reduceItemCount(e) {
    changeItemCount(item, item.count - 1);
    e.stopPropagation();
  }

  function increaseItemCount(e) {
    changeItemCount(item, item.count + 1);
    e.stopPropagation();
  }

  return (
    <div className="ProductCounter">
      <button className="product-counter-btn" onClick={reduceItemCount}>
        -
      </button>
      <div className="counter">{item.count}</div>
      <button className="product-counter-btn" onClick={increaseItemCount}>
        +
      </button>
    </div>
  );
}
