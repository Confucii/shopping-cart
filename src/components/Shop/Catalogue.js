import "./styles/Catalogue.css";
import catalogue from "../shopping";
import Item from "./Item";

export default function Catalogue({
  addItem,
  filter,
  search,
  cart,
  changeItemCount,
}) {
  let filtered = [...catalogue];

  if (filter) {
    filtered = filtered.filter((item) => filter === item.type);
  }

  if (search) {
    const regex = new RegExp(search, "i");
    filtered = filtered.filter((item) => regex.test(item.name));
  }

  return (
    <div className="Catalogue">
      {filtered.map((item) => {
        return (
          <Item
            item={item}
            key={item.id}
            addItem={addItem}
            cart={cart}
            changeItemCount={changeItemCount}
          />
        );
      })}
    </div>
  );
}
