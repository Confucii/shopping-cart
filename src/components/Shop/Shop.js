import { useState } from "react";
import Catalogue from "./Catalogue";
import Filter from "./Filter";
import "./styles/Shop.css";

export default function Shop({ addItem, cart, changeItemCount }) {
  const [filter, setFilter] = useState("");
  const [search, setSearch] = useState("");

  return (
    <div className="Shop">
      <Filter setFilter={setFilter} search={search} setSearch={setSearch} />
      <Catalogue
        addItem={addItem}
        filter={filter}
        search={search}
        cart={cart}
        changeItemCount={changeItemCount}
      />
    </div>
  );
}
