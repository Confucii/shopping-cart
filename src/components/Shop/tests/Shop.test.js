/* eslint-disable testing-library/no-unnecessary-act */
import React from "react";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import Shop from "../Shop";
import userEvent from "@testing-library/user-event";

let cart = [];

function addItem(newItem) {
  const index = cart.findIndex((item) => item.id === newItem.id);
  if (index >= 0) {
    const newCart = [...cart];
    newCart[index].count += 1;
    cart = [...newCart];
  } else {
    cart = [...cart, { id: newItem.id, count: 1 }];
  }
}

function removeItem(deleteItem) {
  cart = cart.filter((item) => item.id !== deleteItem.id);
}

function changeItemCount(changeItem, count) {
  if (count < 1) {
    removeItem(changeItem);
  } else {
    const index = cart.findIndex((item) => item.id === changeItem.id);
    const newCart = [...cart];
    newCart[index].count = count;
    cart = [...newCart];
  }
}

describe("Shop", () => {
  it("Shop is generated", () => {
    const shop = renderer
      .create(
        <BrowserRouter>
          <Shop
            addItem={addItem}
            cart={cart}
            changeItemCount={changeItemCount}
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(shop).toMatchSnapshot();
  });

  it("Products can be added to cart", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );
    const buttons = screen.getAllByRole("button", { name: "Add to cart" });
    userEvent.click(buttons[0]);
    userEvent.click(buttons[3]);
    expect(cart.length).toBe(2);
  });

  it("Product count can be increased", () => {
    cart = [{ id: 1, count: 1 }];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );
    const button = screen.getByRole("button", { name: "+" });

    userEvent.click(button);
    userEvent.click(button);

    expect(cart[0].count).toBe(3);
  });

  it("Products that are already added can be removed", () => {
    cart = [{ id: 1, count: 1 }];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );
    const button = screen.getByRole("button", { name: "-" });

    userEvent.click(button);

    expect(cart.length).toBe(0);
  });

  it("Filter can be applied", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );

    const filterBtn = screen.getByText("Black");
    act(() => {
      userEvent.click(filterBtn);
    });
    const imgs = screen.getAllByRole("img");

    expect(imgs.length).toBe(2);
  });

  it("Filter can be applied and removed", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );

    const filterBtn = screen.getByText("Black");
    act(() => {
      userEvent.click(filterBtn);
      userEvent.click(filterBtn);
    });
    const imgs = screen.getAllByRole("img");

    expect(imgs.length).toBe(8);
  });

  it("Search can be applied", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );

    const searchBox = screen.getByRole("textbox");
    act(() => {
      userEvent.type(searchBox, "Ember");
    });
    const imgs = screen.getAllByRole("img");

    expect(imgs.length).toBe(1);
  });

  it("Search can be cleaned", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );

    const searchBox = screen.getByRole("textbox");
    act(() => {
      userEvent.type(searchBox, "Ember");
      userEvent.clear(searchBox);
    });
    const imgs = screen.getAllByRole("img");

    expect(imgs.length).toBe(8);
  });

  it("Search and filter can be applied", () => {
    cart = [];
    render(
      <BrowserRouter>
        <Shop addItem={addItem} cart={cart} changeItemCount={changeItemCount} />
      </BrowserRouter>
    );

    const filterBtn = screen.getByText("Black");
    const searchBox = screen.getByRole("textbox");

    act(() => {
      userEvent.click(filterBtn);
      userEvent.type(searchBox, "Shining");
    });
    const imgs = screen.getAllByRole("img");

    expect(imgs.length).toBe(1);
  });
});
