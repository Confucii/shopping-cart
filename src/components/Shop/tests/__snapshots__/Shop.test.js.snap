// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop Shop is generated 1`] = `
<div
  className="Shop"
>
  <div
    className="Filter"
  >
    <p
      className="search-name"
    >
      Search
    </p>
    <input
      id="search-input"
      name="search-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p
      className="filter-name"
    >
      Filter
    </p>
    <ul
      className="filter-list"
    >
      <li
        className="filter-item"
        onClick={[Function]}
      >
        Red
      </li>
      <li
        className="filter-item"
        onClick={[Function]}
      >
        Pink
      </li>
      <li
        className="filter-item"
        onClick={[Function]}
      >
        Black
      </li>
      <li
        className="filter-item"
        onClick={[Function]}
      >
        Gold
      </li>
    </ul>
  </div>
  <div
    className="Catalogue"
  >
    <div
      className="Item"
      data-id={0}
      onClick={[Function]}
    >
      <img
        alt="Shining Knight"
        className="item-img"
        src="black_1.png"
      />
      <p>
        Model: 
        Shining Knight
      </p>
      <p>
        Price: 
        30000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={1}
      onClick={[Function]}
    >
      <img
        alt="Ember Destroyer"
        className="item-img"
        src="black_2.png"
      />
      <p>
        Model: 
        Ember Destroyer
      </p>
      <p>
        Price: 
        25000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={2}
      onClick={[Function]}
    >
      <img
        alt="Golden Revenger"
        className="item-img"
        src="gold_1.png"
      />
      <p>
        Model: 
        Golden Revenger
      </p>
      <p>
        Price: 
        35000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={3}
      onClick={[Function]}
    >
      <img
        alt="Radiant Guardian"
        className="item-img"
        src="gold_2.png"
      />
      <p>
        Model: 
        Radiant Guardian
      </p>
      <p>
        Price: 
        40000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={4}
      onClick={[Function]}
    >
      <img
        alt="Crimson Ravager"
        className="item-img"
        src="red_1.png"
      />
      <p>
        Model: 
        Crimson Ravager
      </p>
      <p>
        Price: 
        24000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={5}
      onClick={[Function]}
    >
      <img
        alt="Flush Menace"
        className="item-img"
        src="pink_1.png"
      />
      <p>
        Model: 
        Flush Menace
      </p>
      <p>
        Price: 
        23000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={6}
      onClick={[Function]}
    >
      <img
        alt="Bubblegum Feline"
        className="item-img"
        src="pink_2.png"
      />
      <p>
        Model: 
        Bubblegum Feline
      </p>
      <p>
        Price: 
        20000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
    <div
      className="Item"
      data-id={7}
      onClick={[Function]}
    >
      <img
        alt="Rose Protector"
        className="item-img"
        src="pink_3.png"
      />
      <p>
        Model: 
        Rose Protector
      </p>
      <p>
        Price: 
        15000
         credits
      </p>
      <button
        className="item-btn"
        onClick={[Function]}
      >
        Add to cart
      </button>
    </div>
  </div>
</div>
`;
